packages/shared-metrics/src/dependencies.js

```diff
@@ -269,3 +310,83 @@ function sortByDistance(distances, dependency1, dependency2) {
   }
   return d2 - d1;
 }
+
+/**
+ * Either reads the contents of the main package.json file or returns mainPackageJsonContents if it was cached already
+ * @param {string} mainPackageJsonPath
+ * @param {(err: Error, contents: string) => void} cb
+ */
+function getMainPackageJsonContents(mainPackageJsonPath, cb) {
+  if (mainPackageJsonContents) {
+    cb(null, mainPackageJsonContents);
+  } else {
+    fs.readFile(mainPackageJsonPath, { encoding: 'utf8' }, (readFileErr, contents) => {
+      if (readFileErr) {
+        logger.debug(`Error reading ${mainPackageJsonPath}`, readFileErr);
+        cb(readFileErr, null);
+      } else {
+        mainPackageJsonContents = contents;
+        cb(null, mainPackageJsonContents);
+      }
+    });
+  }
+}
+
+/**
+ * @param {string} mainPackageJsonPath
+ * @param {Object.<string, *>} packages
+ */
+function detectInstalledDevDependencies(mainPackageJsonPath, packages) {
+  if (downstreamConnection) {
+    getMainPackageJsonContents(mainPackageJsonPath, (readFileErr, data) => {
+      if (readFileErr) {
+        logger.debug(`Error reading ${mainPackageJsonPath}`, readFileErr);
+      } else {
+        try {
+          const { devDependencies } = JSON.parse(data.toString());
+
+          if (devDependencies) {
+            const installedDevDeps = [];
+            const devDepKeys = Object.keys(devDependencies);
+
+            for (let i = 0; i < devDepKeys.length; i++) {
+              if (devDepKeys[i] in packages) {
+                installedDevDeps.push(devDepKeys[i]);
```

Comment:
_remark:_ Theoretically, this could lead to false positives. Assuming the application uses package-aaa as a devDependency and package-bbb as a production dependency, but package-bbb also depends transitively on package-aaa -- then this check would detect a devDependency although only production dependencies have been installed. My gut feeling is that this might be a purely academic concern, since applications typically have devDependencies with a larger scope (eslint, mocha, babel, ...) which usually should not be transitive production dependecies.